add convex hull cpu implmentation

musikisomorphie opened the pull request:
@alextp,  I wrote the cpu version of convex hull to address the issue #31067, tfxla version will be implemented later.  Can someone review my PR?

musikisomorphie committed:
add convex hull cpu implmentation

musikisomorphie commented:
Hi, @rmlarsen， @penpornk, how is this PR review going?

penpornk commented:
@musikisomorphie I'll get to this by the end of this week. Sorry for the delay!

musikisomorphie commented:
@penpornk, @rmlarsen hi guys, how is the review going?

gbaned commented:
@musikisomorphie Can you please resolve conflicts? Thanks!

penpornk commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -0,0 +1,151 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
```
Nit: The spacing doesn't look right. Please add newlines around this line.
```suggestion

    http://www.apache.org/licenses/LICENSE-2.0
    
```

musikisomorphie replied:
change it accordingly

penpornk reviewed (changes requested):
@musikisomorphie I'm so sorry for the delay! I mostly have readability comments (for ease of maintenance).

musikisomorphie committed:
Merge pull request #5 from tensorflow/master

update fork

musikisomorphie committed:
Merge branch 'convexhull' into master

musikisomorphie committed:
Merge pull request #6 from musikisomorphie/master

merge master to convexhull

musikisomorphie commented on tensorflow/core/api_def/base_api/api_def_ConvexHull.pbtxt:
```diff
@@ -0,0 +1,26 @@
+  in_arg {
+    name: "input"
+    description: <<END
+3-D with shape `[batch, numbers, dimensions]`.
```
3-D with shape `[batch, num_points, dim_points]`. 
num_points specifies the amount of the points, 
dim_points sepcifies the dimension of the space where the points lie in.

musikisomorphie commented on tensorflow/core/api_def/base_api/api_def_ConvexHull.pbtxt:
```diff
@@ -0,0 +1,26 @@
+  attr {
+    name: "clockwise"
+    description: <<END
+If true, loop through points clockwise. Defaults to false.
```
change to 

> If true, the points are ordered clockwise. 
> If false, the points are oredered counterclockwise.
> Default is false.

musikisomorphie commented on tensorflow/core/api_def/base_api/api_def_ConvexHull.pbtxt:
```diff
@@ -0,0 +1,26 @@
+  }
+  summary: "Obtain the convex hull of given points."
+  description: <<END
+Input can be of different types but output is always float.
```
change to 

>   summary: "Apply sklansky's algorithm to obtain the convex hull of given points."
>   description: <<END
> Returns an array of convex hull points (not indices) for each input batch.
> Input can be of different types but the output is float type.
> END

musikisomorphie commented on tensorflow/core/BUILD:
```diff
@@ -1023,6 +1025,7 @@ cc_library(
         "//tensorflow/core/kernels:collective_ops",
         "//tensorflow/core/kernels:constant_op",
         "//tensorflow/core/kernels:control_flow_ops",
+        "//tensorflow/core/kernels:convex",
```
changed it accordingly

musikisomorphie commented on tensorflow/core/kernels/BUILD:
```diff
@@ -3908,6 +3908,44 @@ tf_cc_test(
     ],
 )
 
+cc_library(
+    name = "convex",
+    deps = [
+        ":convex_hull_op",
+    ],
+)
```
changed it accordingly

musikisomorphie commented on tensorflow/core/kernels/BUILD:
```diff
@@ -3908,6 +3908,44 @@ tf_cc_test(
+    deps = [
+        ":ops_testutil",
+        ":ops_util",
+        ":convex_hull_op",
```
move 

> :convex_hull_op

 to the top of the list

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -0,0 +1,151 @@
+  test::ExpectTensorEqual<float>(expected, *GetOutput(0));
+}
+
+// If two points, return the input points(with differnt order)
```
Yes, it is the case. I tested it on my personal machine.

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -0,0 +1,151 @@
+  test::ExpectTensorEqual<float>(expected, *GetOutput(0));
+}
+
+// If three points not on the same line, return the input points(with differnt
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -0,0 +1,151 @@
+}
+
+// If three points on the same line, return the start and end points (with
+// repitition)
```
change it accordingly

musikisomorphie commented on tensorflow/python/BUILD:
```diff
@@ -3521,6 +3530,7 @@ py_library(
         ":control_flow_grad",
         ":control_flow_ops",
         ":control_flow_util",
+   	":convex_ops",
```
change them accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+                    FastBoundsCheck(input.dim_size(1),
+                                    std::numeric_limits<int64>::max()),
+                errors::InvalidArgument(
+                    "point number must be between 0 and max int64"));
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+};
+
+namespace {
+// This part follows the corresponding opencv implmentation
```
Thanks for the detailed comments, change accordingly.

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+// This part follows the corresponding opencv implmentation
+int64 sklansky(const std::vector<std::vector<float>> &pts_vec,
+               std::vector<int64> &idx_vec, const int64 start, int64 end,
+               const int64 pts_num, const int sign0, const int sign1) {
```
remove 

> pts_num

change to 

> const bool inc_y const bool clockwise

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+int64 sklansky(const std::vector<std::vector<float>> &pts_vec,
+               std::vector<int64> &idx_vec, const int64 start, int64 end,
+               const int64 pts_num, const int sign0, const int sign1) {
+  if (start == end || pts_vec[start] == pts_vec[end]) {
+    idx_vec[0] = start;
+    return 1;
+  }
```
Add comments.

> // If the points vector has one unique point or identical points,
>  // return the point.

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+  while (pnext != end) {
+    // Check the angle of points pprev, pcur, pnext
+    float by = pts_vec[pnext][1] - pts_vec[pcur][1];
+    int by_sign = (by > 0) - (by < 0);
+
+    if (by_sign != sign0) {
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+    int by_sign = (by > 0) - (by < 0);
+
+    if (by_sign != sign0) {
+      float ax = pts_vec[pcur][0] - pts_vec[pprev][0];
+      float bx = pts_vec[pnext][0] - pts_vec[pcur][0];
+      float ay = pts_vec[pcur][1] - pts_vec[pprev][1];
+      // Check if the angle is convex
+      float convex = ay * bx - ax * by;
+      int convex_sign = (convex > 0) - (convex < 0);
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+      float convex = ay * bx - ax * by;
+      int convex_sign = (convex > 0) - (convex < 0);
+
+      if (convex_sign == sign1 && (ax != 0 || ay != 0)) {
```
remove the condition

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+      float convex = ay * bx - ax * by;
+      int convex_sign = (convex > 0) - (convex < 0);
+
+      if (convex_sign == sign1 && (ax != 0 || ay != 0)) {
```
add the comment accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+          pnext += incr;
+          idx_vec[2] = pnext;
+        } else {
+          idx_vec[idx_sz - 2] = pnext;
+          pcur = pprev;
+          pprev = idx_vec[idx_sz - 4];
+          --idx_sz;
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+      }
+    } else {
+      pnext += incr;
+      idx_vec[idx_sz - 1] = pnext;
```
delete the line

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+      idx_vec[idx_sz - 1] = pnext;
+    }
+  }
+  return --idx_sz;
```
delete --

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+void convex_hull(std::vector<std::vector<float>> &pts_vec,
+                 std::vector<std::vector<float>> &out_vec, const bool clockwise,
+                 const int64 pts_num) {
+  auto cmp_x = [](const std::vector<float> &pt0,
+                  const std::vector<float> &pt1) {
+    return pt0[0] != pt1[0] ? pt0[0] < pt1[0] : pt0[1] < pt1[1];
+  };
+  // Sort points based on coordinate x
+  std::sort(pts_vec.begin(), pts_vec.end(), cmp_x);
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+  if (pts_vec[0] == pts_vec[pts_num - 1]) {
+    out_vec[out_idx++] = pts_vec[0];
+  } else {
+    // Upper half of the convex hull
```
add the comments accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+    int u_sign = clockwise ? 1 : -1;
+
+    std::vector<int64> u_vec0(pts_num + 2);
+    int64 u_cnt0 =
+        sklansky(pts_vec, u_vec0, u_st, max_idy, pts_num, -1, u_sign);
+
+    std::vector<int64> u_vec1(pts_num + 2);
+    int64 u_cnt1 = sklansky(pts_vec, u_vec1, pts_num - 1 - u_st, max_idy,
+                            pts_num, -1, -u_sign);
```
change both the upper half and lower half

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+      if (convex_sign == sign1 && (ax != 0 || ay != 0)) {
+        pprev = pcur;
+        pcur = pnext;
+        pnext += incr;
+        idx_vec[idx_sz] = pnext;
```
change pnext to pcur

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+  // Initialize three indices of points
+  int64 pprev = start, pcur = pprev + incr, pnext = pcur + incr;
+  // Initialize the number of convex hull points
+  int64 idx_sz = 3;
+
+  idx_vec[0] = pprev;
+  idx_vec[1] = pcur;
+  idx_vec[2] = pnext;
```
delete the line 

> idx_vec[2] = pnext;

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+        pnext += incr;
+        idx_vec[idx_sz] = pnext;
+        ++idx_sz;
+      } else {
```
add comments accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+          pcur = pnext;
+          idx_vec[1] = pcur;
+          pnext += incr;
+          idx_vec[2] = pnext;
```
delete the line

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+  return --idx_sz;
+}
+
+void convex_hull(std::vector<std::vector<float>> &pts_vec,
```
Document the function similar to function 

> int64 sklansky(...)

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op.cc:
```diff
@@ -0,0 +1,280 @@
+    int64 l_cnt1 = sklansky(pts_vec, l_vec1, pts_num - 1 - l_st, min_idy,
+                            pts_num, 1, -l_sign);
+
+    if (stop_idx >= 0) {
+      int64 check_idx = l_cnt0 > 2
+                            ? l_vec0[1]
+                            : l_cnt0 + l_cnt1 > 2 ? l_vec1[2 - l_cnt0] : -1;
+      if (check_idx == stop_idx ||
+          (check_idx >= 0 && pts_vec[check_idx] == pts_vec[stop_idx])) {
+        l_cnt0 = std::min(l_cnt0, static_cast<int64>(2));
+        l_cnt1 = std::min(l_cnt1, static_cast<int64>(2));
+      }
+    }
```
Add some comments on this part

musikisomorphie committed:
responde to Penporn review

musikisomorphie commented:
Hi @penpornk, thank you for your feedback. Please take a look at my update, which I addressed most of your reviews.

penpornk reviewed (changes requested):
Thank you very much for all the changes! :)
Could you please update the API golden files as well? You can do that by running
```
$ bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
```
and push the changed golden files to the branch. After that, we can tag this PR for API review.

I also have a few more minor formatting/typo nits. Sorry for the inconvenience!

musikisomorphie committed:
Merge pull request #7 from tensorflow/master

update fork

musikisomorphie committed:
Merge pull request #8 from musikisomorphie/master

update convexhull

musikisomorphie committed:
Update tensorflow/core/kernels/convex_hull_op.cc

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

musikisomorphie committed:
Update tensorflow/core/kernels/convex_hull_op.cc

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

musikisomorphie committed:
Update tensorflow/core/kernels/convex_hull_op_test.cc

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

musikisomorphie committed:
Update tensorflow/core/kernels/convex_hull_op.cc

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -60,7 +63,7 @@ TEST_F(ConvexHullOpTest, OnePoint) {
 }
 
-// If two points, return the input points(with differnt order)
+// If two points, return the input points (with differnt order)
```
change it accordingly

musikisomorphie commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -108,7 +111,8 @@ TEST_F(ConvexHullOpTest, ThreePoints_SameLine) {
 // If more than three points,
-// return the convex hull with repititive points padded in the end
+// return the convex hull (with the end point padded
+// to the end of the output array)
```
change it accordingly

musikisomorphie committed:
address penpornks review2

musikisomorphie committed:
Merge pull request #9 from tensorflow/master

update

musikisomorphie committed:
Merge pull request #10 from musikisomorphie/master

update

musikisomorphie commented:
@penpornk, thank you for your feedback. I addressed your reviews. 
However, when I run
 `bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True`
I got the following errors. and I tried compiling tf from scratch several times with different configurations and run the api test, this error remains. Based on the error information, It does not seem to be related to my convexhull implementation. Could you please help me with it? Thank you

```
Running tests under Python 3.7.7: /home/miashan/Applications/miniconda3/envs/torch_test/bin/python3
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 0.0s
I0531 20:55:54.395447 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 0.0s
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibility
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 0.01s
I0531 20:55:54.402081 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 0.01s
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 0.0s
I0531 20:55:54.405161 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 0.0s
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.03s
I0531 20:55:54.431433 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.03s
[  FAILED  ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
WARNING:tensorflow:From /home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/decorator_utils.py:145: GraphKeys.VARIABLES (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.GraphKeys.GLOBAL_VARIABLES` instead.
W0531 20:55:54.438971 140217385260864 deprecation.py:323] From /home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/decorator_utils.py:145: GraphKeys.VARIABLES (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.GraphKeys.GLOBAL_VARIABLES` instead.
WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.backend.get_session is deprecated. Please use tf.compat.v1.keras.backend.get_session instead.

W0531 20:55:54.493530 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.backend.get_session is deprecated. Please use tf.compat.v1.keras.backend.get_session instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.backend.set_session is deprecated. Please use tf.compat.v1.keras.backend.set_session instead.

W0531 20:55:54.493923 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.backend.set_session is deprecated. Please use tf.compat.v1.keras.backend.set_session instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.RandomNormal is deprecated. Please use tf.compat.v1.keras.initializers.RandomNormal instead.

W0531 20:55:54.504908 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.RandomNormal is deprecated. Please use tf.compat.v1.keras.initializers.RandomNormal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.RandomUniform is deprecated. Please use tf.compat.v1.keras.initializers.RandomUniform instead.

W0531 20:55:54.505028 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.RandomUniform is deprecated. Please use tf.compat.v1.keras.initializers.RandomUniform instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.TruncatedNormal is deprecated. Please use tf.compat.v1.keras.initializers.TruncatedNormal instead.

W0531 20:55:54.505102 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.TruncatedNormal is deprecated. Please use tf.compat.v1.keras.initializers.TruncatedNormal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.he_normal is deprecated. Please use tf.compat.v1.keras.initializers.he_normal instead.

W0531 20:55:54.505240 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.he_normal is deprecated. Please use tf.compat.v1.keras.initializers.he_normal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.he_uniform is deprecated. Please use tf.compat.v1.keras.initializers.he_uniform instead.

W0531 20:55:54.505301 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.he_uniform is deprecated. Please use tf.compat.v1.keras.initializers.he_uniform instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.lecun_normal is deprecated. Please use tf.compat.v1.keras.initializers.lecun_normal instead.

W0531 20:55:54.505368 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.lecun_normal is deprecated. Please use tf.compat.v1.keras.initializers.lecun_normal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.lecun_uniform is deprecated. Please use tf.compat.v1.keras.initializers.lecun_uniform instead.

W0531 20:55:54.505424 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.lecun_uniform is deprecated. Please use tf.compat.v1.keras.initializers.lecun_uniform instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.normal is deprecated. Please use tf.compat.v1.keras.initializers.normal instead.

W0531 20:55:54.505477 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.normal is deprecated. Please use tf.compat.v1.keras.initializers.normal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.random_normal is deprecated. Please use tf.compat.v1.keras.initializers.random_normal instead.

W0531 20:55:54.505549 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.random_normal is deprecated. Please use tf.compat.v1.keras.initializers.random_normal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.random_uniform is deprecated. Please use tf.compat.v1.keras.initializers.random_uniform instead.

W0531 20:55:54.505603 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.random_uniform is deprecated. Please use tf.compat.v1.keras.initializers.random_uniform instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.truncated_normal is deprecated. Please use tf.compat.v1.keras.initializers.truncated_normal instead.

W0531 20:55:54.505664 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.truncated_normal is deprecated. Please use tf.compat.v1.keras.initializers.truncated_normal instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.uniform is deprecated. Please use tf.compat.v1.keras.initializers.uniform instead.

W0531 20:55:54.505716 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.initializers.uniform is deprecated. Please use tf.compat.v1.keras.initializers.uniform instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.CuDNNGRU is deprecated. Please use tf.compat.v1.keras.layers.CuDNNGRU instead.

W0531 20:55:54.510276 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.CuDNNGRU is deprecated. Please use tf.compat.v1.keras.layers.CuDNNGRU instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.CuDNNLSTM is deprecated. Please use tf.compat.v1.keras.layers.CuDNNLSTM instead.

W0531 20:55:54.510368 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.CuDNNLSTM is deprecated. Please use tf.compat.v1.keras.layers.CuDNNLSTM instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.disable_v2_dtype_behavior is deprecated. Please use tf.compat.v1.keras.layers.disable_v2_dtype_behavior instead.

W0531 20:55:54.510708 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.disable_v2_dtype_behavior is deprecated. Please use tf.compat.v1.keras.layers.disable_v2_dtype_behavior instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.enable_v2_dtype_behavior is deprecated. Please use tf.compat.v1.keras.layers.enable_v2_dtype_behavior instead.

W0531 20:55:54.510774 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.layers.enable_v2_dtype_behavior is deprecated. Please use tf.compat.v1.keras.layers.enable_v2_dtype_behavior instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.losses.cosine is deprecated. Please use tf.keras.losses.cosine_similarity instead.

W0531 20:55:54.585864 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.losses.cosine is deprecated. Please use tf.keras.losses.cosine_similarity instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.losses.cosine_proximity is deprecated. Please use tf.keras.losses.cosine_similarity instead.

W0531 20:55:54.586018 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.losses.cosine_proximity is deprecated. Please use tf.keras.losses.cosine_similarity instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.metrics.cosine is deprecated. Please use tf.keras.losses.cosine_similarity instead.

W0531 20:55:54.589288 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.metrics.cosine is deprecated. Please use tf.keras.losses.cosine_similarity instead.

WARNING:tensorflow:From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.metrics.cosine_proximity is deprecated. Please use tf.keras.losses.cosine_similarity instead.

W0531 20:55:54.589395 140217385260864 module_wrapper.py:138] From /home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/inspect.py:341: The name tf.keras.metrics.cosine_proximity is deprecated. Please use tf.keras.losses.cosine_similarity instead.

INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.27s
I0531 20:55:54.703651 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.27s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.23s
I0531 20:55:54.936919 140217385260864 test_util.py:1970] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.23s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
======================================================================
ERROR: testAPIBackwardsCompatibility (__main__.ApiCompatibilityTest)
testAPIBackwardsCompatibility (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 410, in testAPIBackwardsCompatibility
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 359, in _checkBackwardsCompatibility
    traverse.traverse(root, public_api_visitor)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 104, in traverse
    _traverse_internal(root, visit, [], '')
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 54, in _traverse_internal
    visit(path, root, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/public_api.py", line 140, in __call__
    self._visitor(path, parent, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 242, in __call__
    _AddMember(name, child, module_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 207, in _AddMember
    _, member_obj = tf_decorator.unwrap(member_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 219, in unwrap
    elif _has_tf_decorator_attr(cur):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 124, in _has_tf_decorator_attr
    hasattr(obj, '_tf_decorator') and
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/__init__.py", line 10, in <module>
    from tensorflow_estimator._api.v1 import estimator
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/_api/v1/estimator/__init__.py", line 10, in <module>
    from tensorflow_estimator._api.v1.estimator import experimental
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/_api/v1/estimator/experimental/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.canned.dnn import dnn_logit_fn_builder
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py", line 23, in <module>
    from tensorflow.python.feature_column import dense_features
ImportError: cannot import name 'dense_features' from 'tensorflow.python.feature_column' (/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/feature_column/__init__.py)

======================================================================
ERROR: testAPIBackwardsCompatibilityV1 (__main__.ApiCompatibilityTest)
testAPIBackwardsCompatibilityV1 (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 426, in testAPIBackwardsCompatibilityV1
    omit_golden_symbols_map={'tensorflow': ['pywrap_tensorflow']})
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 359, in _checkBackwardsCompatibility
    traverse.traverse(root, public_api_visitor)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 104, in traverse
    _traverse_internal(root, visit, [], '')
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 54, in _traverse_internal
    visit(path, root, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/public_api.py", line 140, in __call__
    self._visitor(path, parent, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 242, in __call__
    _AddMember(name, child, module_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 207, in _AddMember
    _, member_obj = tf_decorator.unwrap(member_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 219, in unwrap
    elif _has_tf_decorator_attr(cur):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 124, in _has_tf_decorator_attr
    hasattr(obj, '_tf_decorator') and
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1 import estimator
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1.estimator import experimental
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/experimental/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.canned.dnn import dnn_logit_fn_builder
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py", line 23, in <module>
    from tensorflow.python.feature_column import dense_features
ImportError: cannot import name 'dense_features' from 'tensorflow.python.feature_column' (/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/feature_column/__init__.py)

======================================================================
ERROR: testAPIBackwardsCompatibilityV2 (__main__.ApiCompatibilityTest)
testAPIBackwardsCompatibilityV2 (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 443, in testAPIBackwardsCompatibilityV2
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 359, in _checkBackwardsCompatibility
    traverse.traverse(root, public_api_visitor)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 104, in traverse
    _traverse_internal(root, visit, [], '')
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 54, in _traverse_internal
    visit(path, root, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/public_api.py", line 140, in __call__
    self._visitor(path, parent, children)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 242, in __call__
    _AddMember(name, child, module_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/lib/python_object_to_proto_visitor.py", line 207, in _AddMember
    _, member_obj = tf_decorator.unwrap(member_obj)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 219, in unwrap
    elif _has_tf_decorator_attr(cur):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 124, in _has_tf_decorator_attr
    hasattr(obj, '_tf_decorator') and
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1 import estimator
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1.estimator import experimental
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/experimental/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.canned.dnn import dnn_logit_fn_builder
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py", line 23, in <module>
    from tensorflow.python.feature_column import dense_features
ImportError: cannot import name 'dense_features' from 'tensorflow.python.feature_column' (/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/feature_column/__init__.py)

======================================================================
ERROR: testNoSubclassOfMessage (__main__.ApiCompatibilityTest)
testNoSubclassOfMessage (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 317, in testNoSubclassOfMessage
    traverse.traverse(tf, visitor)
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 104, in traverse
    _traverse_internal(root, visit, [], '')
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/common/traverse.py", line 58, in _traverse_internal
    child) and child.__name__ in sys.builtin_module_names:
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_inspect.py", line 397, in ismodule
    return _inspect.ismodule(tf_decorator.unwrap(object)[1])
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 219, in unwrap
    elif _has_tf_decorator_attr(cur):
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/tf_decorator.py", line 124, in _has_tf_decorator_attr
    hasattr(obj, '_tf_decorator') and
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1 import estimator
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.api._v1.estimator import experimental
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/api/_v1/estimator/experimental/__init__.py", line 10, in <module>
    from tensorflow_estimator.python.estimator.canned.dnn import dnn_logit_fn_builder
  File "/home/miashan/Applications/miniconda3/envs/torch_test/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/canned/dnn.py", line 23, in <module>
    from tensorflow.python.feature_column import dense_features
ImportError: cannot import name 'dense_features' from 'tensorflow.python.feature_column' (/home/miashan/.cache/bazel/_bazel_miashan/7f08cea0e8cec4ae4ded56090269931f/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/feature_column/__init__.py)

----------------------------------------------------------------------
Ran 7 tests in 0.546s

FAILED (errors=4, skipped=1)
```

penpornk commented on tensorflow/core/kernels/convex_hull_op_test.cc:
```diff
@@ -153,4 +153,5 @@ TEST_F(ConvexHullOpTest, Error_InputDimMustBe2) {
 }  // namespace
-}  // namespace tensorflow
\ No newline at end of file
+
```
One last minor nit please (since we need another commit for the golden files anyway). Ends of namespaces don't need to be separated by a blank line. Would you mind removing this line? (But please keep the newline at the end of file.)
```c++
}  // namespace
}  // namespace tensorflow
```

musikisomorphie replied:
removed the redundant line

penpornk reviewed (changes requested):
@musikisomorphie Thank you for the fixes!

Yes, the error is unrelated to your PR (also happens with the TF master branch). It's a TF estimator version problem. Switching to a nightly package should fix it.
```
pip3 uninstall tensorflow_estimator
pip3 install tf-estimator-nightly
bazel run //tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
```

Let me know if this works.

musikisomorphie committed:
Merge pull request #11 from tensorflow/master

update

musikisomorphie committed:
Merge pull request #12 from musikisomorphie/master

update convexhull branch

musikisomorphie committed:
update goldens, remove redundant line

add new line at the end

musikisomorphie commented:
@pengchongjin, thanks for the answer. I updated the API golden files, fixed the empty line issues. You could take a look.

penpornk reviewed (approved):
Thank you very much again for the PR! I'll tag this for API review.

martinwicke commented:
On the face of it, this looks like tensorflow/graphics might be the best fit. 

@sofienbouaziz WDYT?

musikisomorphie commented:
> On the face of it, this looks like tensorflow/graphics might be the best fit.
> 
> @sofienbouaziz WDYT?

This function is originally implemented in opencv, I am not sure if we should migrate it to tfg.

musikisomorphie commented:
hi @penpornk, @martinwicke guys, what is the next step of this PR?

penpornk commented:
@musikisomorphie I think we are waiting to hear from @sofienbouaziz and then @martinwicke will decide the next step. Sorry for the inconvenience and thank you for your patience!

sofienbouaziz commented:
Sorry for the late reply, we have been discussing this PR quite a bit internally. In general I think having a convex hull functionality is a great match for TensorFlow Graphics. Though I have a few concerns:
1) It implements Sklansky's algorithm (same than OpenCV) which is fairly simple but to the best of my knowledge can in some cases result in wrong outputs
2) The current PR only provides a C++ implementation (and no gradient as it is not differentiable out of the box). I am wondering what do we gain over using tf.numpy_function with OpenCV (or SciPy) as we will need to maintain the code and make sure it is bug free. 
3) There is a mention of a future tfxla implementation which seems quite hard to implement due to the dynamic shapes. It seems also not easy to parallelize.

musikisomorphie commented:
> Sorry for the late reply, we have been discussing this PR quite a bit internally. In general I think having a convex hull functionality is a great match for TensorFlow Graphics. This is a great PR, though I have a few concerns:
> 
> 1. It implements Sklansky's algorithm (same than OpenCV) which is fairly simple but to the best of my knowledge can in some cases result in wrong outputs
> 2. The current PR only provides a C++ implementation (and no gradient as it is not differentiable out of the box). I am wondering what do we gain over using tf.numpy_function with OpenCV (or SciPy) as we will need to maintain the code and make sure it is bug free.
> 3. There is a mention of a future tfxla implementation which seems quite hard to implement due to the dynamic shapes. It seems also not easy to parallelize.

@sofienbouaziz, thanks for the feedback. I am open to different options, just let me know what the final decision is.

sofienbouaziz commented:
@musikisomorphie I discussed a bit with @taiya and he mentioned that maybe a better option could be to provide a frontend for QHull's C API instead of reimplementing. The main question for me is what do we get over using tf.numpy_function and directly calling scipy which seem to use QHull internally? Maintaining C++ code has a non negligible cost and I am trying to understand the opportunity cost tradeoff.

alextp commented:
@sofienbouaziz if you can use scipy's API that would of course be lovely. Some caveats / comments:

 1. you only need to expose it as a TF op if you want to use it in the middle of a TF model; using it before or after the part you want to put inside a tf.function or differentiate doesn't require any work as you can transparently convert back and forth between TF tensors and numpy ndarrays
 2. if you do want to use tf.numpy_function bear in mind that it won't be serializable in a savedmodel and it won't be differentiable; implementing convex hull as a TF op in C++ will allow both things to be supported
 3. an option is maybe to just call some C++ implementation somewhere?

I also like the idea of writing an implementation of a simple graham scan algorithm using tf.sort and tf.TensorArray to implement the stack and a tf.while_loop to drive the loop; this would perform abysmally when interpreted but if we can XLA compile it we can probably get pretty nice performance (most likely close to but not exceeding a straightforward hand coded implementation like you have here but with the potential of TPU support).

gbaned commented:
@musikisomorphie, @sofienbouaziz Any update on this PR? Please. Thanks!

musikisomorphie commented:
> @musikisomorphie, @sofienbouaziz Any update on this PR? Please. Thanks!



> @musikisomorphie, @sofienbouaziz Any update on this PR? Please. Thanks!

hi guys @sofienbouaziz @alextp , I feel most likely you are going for an alternative approach rather than my implementation. I guess I am going to focus on something else instead.  feel free to take my implementation as a reference.

taiya commented:
Apologies, unless we know this would be better than what is already
available in python... it is a hard sell.
Would you mind closing the PR? THanks

musikisomorphie commented:
> Apologies, unless we know this would be better than what is already available in python... it is a hard sell. Would you mind closing the PR? THanks

Sure, no problem.
